File: src/tree_parity_machine/neuron/Neuron.java
package tree_parity_machine.neuron;

import learning_algorithm.LearningParadigm;
import tree_parity_machine.NeuralNetException;

import java.util.Arrays;

public abstract class Neuron {

    protected int output;
    protected int inputs;
    protected double[] weights;
    protected int leftBound;
    protected int rightBound;
    protected LearningParadigm paradigm;

    public abstract void init();

    public abstract void changeWeights(double[] input, int outputTPM);

    public abstract int getOutput(double[] input) throws NeuralNetException;

    public double[] getWeights() {
        return weights;
    }

    public LearningParadigm getParadigm() {
        return paradigm;
    }

    public void setParadigm(LearningParadigm paradigm) {
        this.paradigm = paradigm;
    }

    @Override
    public String toString() {
        return "Neuron{" +
                "inputs=" + inputs +
                ", weights=" + Arrays.toString(weights) +
                '}';
    }
}

File: src/main/kotlin/com/jvvladimir/machine/tree_machine/neuron/Neuron.kt
package com.jvvladimir.machine.tree_machine.neuron

import com.jvvladimir.machine.learning.LearningParadigm

abstract class Neuron {
    protected var output = 0
    protected var inputs = 0
    lateinit var weights: IntArray
        protected set
    protected var leftBound = 0
    protected var rightBound = 0
    lateinit var paradigm: LearningParadigm

    abstract fun init()
    abstract fun changeWeights(input: IntArray, outputTPM: Int)
    abstract fun getOutput(input: IntArray): Int
    override fun toString(): String {
        return "Neuron{" +
                "output=" + output +
                ", inputs=" + inputs +
                ", weights=" + weights.contentToString() +
                ", leftBound=" + leftBound +
                ", rightBound=" + rightBound +
                ", paradigm=" + paradigm +
                '}'
    }
}